### PR TITLE
Harmonize user data across persistence strategies

### PR DESCRIPTION
**Pull request by binh-dam-ibigroup**
This PR exposes the logged-in user saved places and recent searches in `DefaultMainPanel` and `BatchRoutingPanel`, and in the `LocationField` drop-downs in the main panel.

Behind the scenes:
- Some new place components are introduced:
  - `<Place>` is the generic rendering component, takes a `name`, `details` and `onClick`, `onView`, `onDelete` event handlers. Variants of this component are used in the main panel and in the user's settings page.
- Only the saved/favorite locations and recent trip requests for the persistence strategy defined in `config.yml` are shown. Note: recent trip requests for an OTP Middleware user cannot be deleted.
- ~Only the first OTP web request of a batch request is logged for middleware users, to avoid repeat entries in recent searches.~ => No longer applies, a `batchId` must now be sent on every trip request to OTP after the change from https://github.com/ibi-group/otp-middleware/pull/152.
- "Starring" favorite stops from the stop viewer is disabled for logged-in middleware users.
- The user-related reducer code is moved to `create-user-reducer`.
- i18n is implemented, except for content rendered by `LocationField`.

This PR supersedes #283.

Screenshot for the `otp_middleware` persistence strategy:

image

### CODE REVIEW

**Comment by binh-dam-ibigroup** [comment]
I am reverting this to draft. There are quite a few things that broke and that I subsequently fixed, so it makes sense to not request multiple reviews on changes to the same code.

**Comment by binh-dam-ibigroup** [comment]
I'm hitting the point of diminishing returns for this PR, but this should be ready for review now.

**Review by landonreed — CHANGES_REQUESTED** [review]
I haven't fully reviewed this, but based on the PR description, I think there's probably some room for simplifying things a bit. When we discussed using this PR to harmonize the cached user settings with the middleware-persisted settings, my thinking was that we would only rely on a single data model for all of these settings. This would remove the need for a `CachedPlace` and a `SavedPlace` and (hopefully) remove the need for some of the duplicated actions and reducer logic.

Was there an overwhelming reason not to do more harmonization? If this requires breaking changes to the `UserSettings` component or others, I'm happy to entertain those (especially because that component is not in production anywhere to my knowledge).

So, my recommendation would be to not keep a separate `localUser` and `loggedInUser` and reduce the use of components specific to a persistence strategy.

File: __tests__/reducers/create-user-reducer.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import {getUserInitialState} from '../../lib/reducers/create-user-reducer'
+import {restoreDateNowBehavior, setDefaultTestTime} from '../test-utils'
+
+describe('lib > reducers > create-user-reducer', () => {
+  afterEach(restoreDateNowBehavior)
+
+  it('should be able to create the initial state', () => {
+    setDefaultTestTime()
+    expect(getUserInitialState({}, {})).toMatchSnapshot()
```
Comment (landonreed) [inline __tests__/reducers/create-user-reducer.js:9]: I don't think this method `getUserInitialState` has two args.

**Review by evansiroky — CHANGES_REQUESTED** [review]
Is this still ready for review? I seem to remember we chatted about some kind of change to localStorage, so I was wondering if there was going to be some kind of work on that included in this PR. Either way there are now merge conflicts, so please reassign once the merge conflicts are resolved and this is ready for review again.

**Comment by binh-dam-ibigroup** [comment]
> Is this still ready for review? I seem to remember we chatted about some kind of change to localStorage, so I was wondering if there was going to be some kind of work on that included in this PR. Either way there are now merge conflicts, so please reassign once the merge conflicts are resolved and this is ready for review again.

@evansiroky is right, I'm adding the BLOCKED label pending internal decision on the `localStorage` strategy.

**Comment by landonreed** [comment]
@binh-dam-ibigroup, I've refactored a fair amount of stuff here, but there are still some failing tests. This simplified component structure should be more maintainable over the long term and I think we should try to just have the localstorage user object match the middleware model if possible. Feel free to take this and run with it!

File: __tests__/reducers/create-user-reducer.js (hunk 1 above)
Comment (binh-dam-ibigroup) [reply __tests__/reducers/create-user-reducer.js]: Removed in https://github.com/opentripplanner/otp-react-redux/pull/329/commits/b7b608d2c9c8a57b9bbcbe28b81a59ef0b41a77e.

File: lib/components/form/user-settings.js [hunk 2]
```diff
@@ -1,308 +1,424 @@
...
-import Icon from '../util/icon'
+import { UnpaddedList } from './styled'
 
-const { formatStoredPlaceName, getDetailText, matchLatLon } = coreUtils.map
-const { summarizeQuery } = coreUtils.query
+const { matchLatLon } = coreUtils.map
+const { hasTransit, toSentenceCase } = coreUtils.itinerary
+/**
+ * Version of summarizeQuery and helper function that supports i18n.
+ * FIXME: replace when the original does support i18n.
```
Comment (miles-grant-ibigroup) [inline lib/components/form/user-settings.js:32]: Starting to feel like this PR is cursed, but do you want to block until https://github.com/opentripplanner/otp-ui/pull/330 is merged?

File: lib/components/form/user-settings.js [hunk 3]
```diff
@@ -1,308 +1,424 @@
...
 class UserSettings extends Component {
   _disableTracking = () => {
-    const { intl, toggleTracking, user } = this.props
-    if (!user.trackRecent) return
-    const hasRecents = user.recentPlaces.length > 0 || user.recentSearches.length > 0
+    const { intl, localUser, toggleTracking } = this.props
+    const { recentPlaces, recentSearches, storeTripHistory } = localUser
+    if (!storeTripHistory) return
+    const hasRecents =
+      (recentPlaces && recentPlaces.length > 0) ||
```
Comment (miles-grant-ibigroup) [inline lib/components/form/user-settings.js:86]: Looks like we can use optional chaining here!

Screen Shot 2022-01-28 at 8 03 11 AM

File: lib/components/map/connected-endpoints-overlay.js [hunk 4]
```diff
@@ -30,11 +41,12 @@ const mapStateToProps = (state, ownProps) => {
   const { from, to } = query
   // Intermediate places doesn't trigger a re-plan, so for now default to
   // current query. FIXME: Determine with TriMet if this is desired behavior.
-  const places = state.otp.currentQuery.intermediatePlaces.filter(p => p)
+  const places = state.otp.currentQuery.intermediatePlaces.filter((p) => p)
```
Comment (miles-grant-ibigroup) [inline lib/components/map/connected-endpoints-overlay.js:44]: What is this filter doing?

File: lib/components/user/places/place.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  const iconSize = largeIcon && '2x'
+  return (
+    <Container className={className}>
+      <LinkContainerWithQuery
+        // Don't highlight component if 'to' is null.
+        activeClassName=""
+        to={to}
+      >
+        <PlaceButton onClick={onClick} title={title}>
+          {largeIcon && <Icon size="2x" type={icon} />}
```
Comment (miles-grant-ibigroup) [inline lib/components/user/places/place.js:70]: Could we be using `iconSize` here?

File: lib/components/user/places/place.js [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+`
+export const PlaceButton = styled(Button)`
+  background: none;
+  flex: 1 0 0;
+  overflow: hidden;
+  text-align: left;
+  text-overflow: ellipsis;
+`
+
+export const PlaceDetail = styled.span``
```
Comment (miles-grant-ibigroup) [inline lib/components/user/places/place.js:22]: Why did you decide to put these here instead of in the component's `styled.js`?

**Review by miles-grant-ibigroup — CHANGES_REQUESTED** [review]
Thank you so much for wrestling this thing to run again. Very excited to try it out! And then I'll be able to do a more proper review

File: lib/components/map/connected-endpoints-overlay.js (hunk 4 above)
Comment (binh-dam-ibigroup) [reply lib/components/map/connected-endpoints-overlay.js:44]: I think this filter keeps the non-null intermediate places.

File: lib/components/form/user-settings.js (hunk 2 above)
Comment (binh-dam-ibigroup) [reply lib/components/form/user-settings.js:32]: I think they can be done independently.

File: lib/util/user.js [hunk 7]
```diff
@@ -43,37 +57,165 @@ export const PLACE_TYPES = {
...
+
+/**
+ * Obtains the saved and recent locations based on the redux state.
+ */
+export function getUserLocations(state) {
+  let saved = []
+  let recent = []
+
+  const { persistence } = state.otp.config
+  const { localUser, loggedInUser } = state.user
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/util/user.js:158]: Thanks... fixed in https://github.com/opentripplanner/otp-react-redux/pull/329/commits/d2ce58fe60a9a6f465297efb3fc507cb6f3b14a0.

File: lib/components/user/places/place.js (hunk 6 above)
Comment (binh-dam-ibigroup) [reply lib/components/user/places/place.js:22]: To avoid a circular reference between styled.js and place.js

Comment (binh-dam-ibigroup) [reply lib/components/user/places/place.js:22]: That way also, styled.js defines different flavors of the `<Place>` component and its internal elements.

**Comment by binh-dam-ibigroup** [comment]
Ready for review again, hopefully!

File: lib/components/user/places/place.js (hunk 5 above)
Comment (binh-dam-ibigroup) [reply lib/components/user/places/place.js]: Changed in https://github.com/opentripplanner/otp-react-redux/pull/329/commits/5f9795a0b3f3911ee4173702d7a4c11a97bcb022.

File: lib/components/form/user-settings.js (hunk 3 above)
Comment (binh-dam-ibigroup) [reply lib/components/form/user-settings.js]: Changed in https://github.com/opentripplanner/otp-react-redux/pull/329/commits/5f9795a0b3f3911ee4173702d7a4c11a97bcb022.

File: lib/actions/user.js [hunk 8]
```diff
@@ -109,6 +133,88 @@ export function fetchAuth0Token(auth0, intl) {
...
+
+    const { data: requests, status: emptyStatus } = await secureFetch(
+      `${requestUrl}&limit=0`,
+      accessToken,
+      apiKey,
+      'GET'
+    )
+
+    // FIXME: Revert to 10 after taking care of duplicates in the middleware.
+    const DEFAULT_LIMIT = 50
```
Comment (miles-grant-ibigroup) [inline lib/actions/user.js:192]: Are the duplicates fixed?

**Review by miles-grant-ibigroup — APPROVED** [review]
There doesn't seem to be any new documentation so I couldn't find out which flags in the config file enable all the new features, but from what I looked at things look like they work great! So happy that this is finally ready to go!

**Comment by binh-dam-ibigroup** [comment]
> There doesn't seem to be any new documentation so I couldn't find out which flags in the config file enable all the new features, but from what I looked at things look like they work great! So happy that this is finally ready to go!

There are no new config settings, as this PR unifies the user settings for the `localStorage` (for non-middleware configs) and the `otp_middleware` persistence strategies.

File: lib/actions/user.js (hunk 8 above)
Comment (binh-dam-ibigroup) [reply lib/actions/user.js:192]: I think we would have to review how trip requests are recorded.

**Comment by daniel-heppner-ibigroup** [comment]
I'm getting some errors while making a new account running FDOT. Are these related to this PR? 
On this URL: http://localhost:9966/#/account/create/terms
image

File: lib/components/user/places/place.js [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+              {!largeIcon && <Icon type={icon} />}
+              <PlaceName>{mainText}</PlaceName>
+            </PlaceText>
+            <PlaceDetail className="place-detail">{detailText}</PlaceDetail>
+          </PlaceContent>
+        </PlaceButton>
+      </LinkContainerWithQuery>
+
+      {/* Action buttons. If none, render a placeholder. */}
+      {!onView && !onDelete && <ActionButtonPlaceholder />}
```
Comment (daniel-heppner-ibigroup) [inline lib/components/user/places/place.js:87]: Would using `or`s work here? 
`onView || onDelete || <ActionButtonPlaceholder />`

**Review by daniel-heppner-ibigroup — APPROVED** [review]

File: lib/util/user.js [hunk 10]
```diff
@@ -43,37 +57,165 @@ export const PLACE_TYPES = {
...
+/**
+ * Convert a LocationField entry to a persisted user savedLocations:
+ * - The icon for "Work" places is changed to 'briefcase',
+ * - The address attribute is filled with the 'name' if available.
+ */
+export function convertToPlace(location) {
+  const { icon, id, lat, lon, name, type } = location
+  return {
+    address: name,
+    icon: isWork(location) ? 'briefcase' : icon,
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/util/user.js:124]: With the `localStorage` strategy, the previous code would derive the icon type from the place type, in that case 'work', that doesn't correspond to a font-awesome icon. Let me know if I should remove it.

File: lib/components/user/places/place.js (hunk 9 above)
Comment (binh-dam-ibigroup) [reply lib/components/user/places/place.js:87]: I will not implement. Your suggestion renders correctly, but it's a little harder to read, and if either `onView` or `onDelete` are truthy, this statement results in a react warning (you are not supposed to render functions).

File: lib/util/user.js [hunk 11]
```diff
@@ -43,37 +57,165 @@ export const PLACE_TYPES = {
...
 }
+
+/**
+ * Returns the persistence.strategy string if the following apply in config.yml, null otherwise:
+ * - persistence is defined,
+ * - persistence.enabled is true,
+ * - persistence.strategy is defined.
+ */
+function getPersistenceStrategy(persistence) {
+  return persistence && persistence.enabled && persistence.strategy
```
[the remark this answers is not in the document]
Comment (binh-dam-ibigroup) [reply lib/util/user.js]: I have shortened the statement using `?.` in https://github.com/opentripplanner/otp-react-redux/pull/329/commits/1b6a0254de19361c10191e5d8c59aa9dd801fdf5

**Pull request merged**